auth/oidc.go

```diff
@@ -248,88 +327,176 @@ func (op *OIDCProvider) InitOIDCClient() error {
 	if err != nil {
 		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
-		return nil, err
+		return ProviderMetadata{}, 0, false, err
 	}
-	defer func() { _ = resp.Body.Close() }()
+
+	defer func() {
+		_ = resp.Body.Close()
+	}()
 	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("%s: %s", resp.Status, resp.Body)
+		return ProviderMetadata{}, 0, false, fmt.Errorf("unsuccessful response: %v", err)
+	}
+
+	ttl, _, err = cacheable(resp.Header)
+	if err != nil {
+		return ProviderMetadata{}, 0, false, err
+	}
+
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return ProviderMetadata{}, 0, false, err
 	}
-	var metadata ProviderMetadata
-	if err := base.JSONDecoder(resp.Body).Decode(&metadata); err != nil {
-		base.Debugf(base.KeyAuth, "Error parsing body %s: %v", base.UD(discoveryURL), err)
-		return nil, err
+
+	if err := base.JSONUnmarshal(bodyBytes, &metadata); err != nil {
+		base.Debugf(base.KeyAuth, "Error parsing body during discovery sync: %v", err)
+		return ProviderMetadata{}, 0, false, err
+	}
+
+	if reflect.DeepEqual(op.metadata, metadata) {
+		base.Debugf(base.KeyAuth, "No change in discovery config detected at this time, next sync will be after %v", ttl)
+		return metadata, ttl, false, nil
 	}
 
 	if metadata.Issuer != op.Issuer {
-		return nil, fmt.Errorf("oidc: issuer did not match the issuer returned by provider, expected %q got %q", op.Issuer, metadata.Issuer)
+		return ProviderMetadata{}, 0, false, fmt.Errorf("oidc: issuer did not match the issuer returned by provider, expected %q got %q", op.Issuer, metadata.Issuer)
 	}
 
 	base.Debugf(base.KeyAuth, "Returning config: %v", base.UD(metadata))
-	return &metadata, nil
+	return metadata, ttl, true, nil
+}
+
+// runDiscoverySync runs the discovery sync by fetching the provider metadata by calling
+// FetchCustomProviderConfig. If FetchCustomProviderConfig detects any change in provider
+// config, it sets the verifier, OAuth endpoints and returns the time interval between now
+// and next discovery sync.
+func (op *OIDCProvider) runDiscoverySync(discoveryURL string) (time.Duration, error) {
+	metadata, ttl, refresh, err := op.FetchCustomProviderConfig(discoveryURL)
+	if err != nil {
+		return 0, err
+	}
+	if refresh && op.getDiscoveryScheme() == nonstandardDiscovery {
+		verifier := op.GenerateVerifier(&metadata, context.Background())
+		op.client.SetVerifier(verifier)
+		op.client.SetEndpoint(metadata.Endpoint())
+		op.metadata = metadata
+	}
+	if refresh && op.getDiscoveryScheme() == standardDiscovery {
+		metadata, verifier, err := op.DiscoverConfig()
+		if err != nil {
+			return 0, pkgerrors.Wrap(err, ErrMsgUnableToDiscoverConfig)
+		}
+		verifier = op.GenerateVerifier(&metadata, context.Background())
+		if verifier == nil {
+			return 0, pkgerrors.Wrap(err, ErrMsgUnableToGenerateVerifier)
+		}
+		op.client.SetVerifier(verifier)
+		op.client.SetEndpoint(metadata.Endpoint())
```

Comment:
Verifier and Endpoint appear to always be updated together.  Can we combine these updates into a single function, so that we don't need to acquire the mutex twice, and don't have to worry about what happens if a client retrieves them when the verifier has been updated but the endpoint hasn't?